ops/main.py

```diff
@@ -42,6 +43,134 @@
 logger = logging.getLogger()
 
 
+@dataclasses.dataclass(frozen=True)
```

Comment:
`init=False` would make that worse, no? Because then it will have `object.__init__` that takes no arguments, so you'd always have the defaults. Plus I think this would break the creation of the object in the classmethod, no?

The proper fix, imo, is that anything that isn't optional should either be fixed in `__post_init__` or (better) should have a `dataclasses.field` with an appropriate default factory.

Reply:
I thought we would like people to instantiate partial _JujuContext in tests.

At least in the tests for ops itself.

Reply:
> I thought we would like people to instantiate partial _JujuContext in tests.
> 
> At least in the tests for ops itself.

In the test, we can create a mocked dictionary then `_JujuContext.from_dict(d)`, it's even easier than monkey patching, and I think this is why Ben suggested passing a dict to `from_dict` in the first place.